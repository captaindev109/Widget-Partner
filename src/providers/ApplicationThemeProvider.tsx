import { ThemeProvider, DefaultTheme } from 'styled-components';
import GlobalStyles from '@/styles/Global';
import React from 'react';
import { ChildrenReturnType } from '@/types/types';

interface ApplicationThemeProviderProps {
  children?: ChildrenReturnType;
  theme: DefaultTheme;
}

const ApplicationThemeProvider = ({ children, theme }: ApplicationThemeProviderProps) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};

export default ApplicationThemeProvider;
