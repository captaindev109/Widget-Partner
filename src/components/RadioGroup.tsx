import React, { FC, useEffect, useRef } from 'react';
import Card from '@/generics/Card';
import styled from 'styled-components';
import Label from '@/generics/Label';
import Tooltip from './Tooltip';
import Input from '@/generics/Input';
import { CreateSVGComponent } from './SVGIcons/Index';
import { animateOpacity, ANIMATION_DURATION } from '@/utils/animationUtils';
import Text from '@/generics/Text';
import { RadioGroupProps } from '@/types/types';

const TopRow = styled(Card)`
  display: flex;
  align-items: center;

  & > * {
    margin-left: 10px;
  }

  margin-bottom: 8px;
`;

const StyledIcon = styled(Card)<{ $animate: boolean; error?: boolean }>`
  color: ${(props) => (props.error ? props.theme.color.red600 : props.theme.color.gray700)};
  transition: all ${ANIMATION_DURATION} ease;

  ${(props) => props.$animate && animateOpacity()}
`;

const HelperText = styled(Text)<{ $animate: boolean; error?: boolean }>`
  color: ${(props) => (props.error ? props.theme.color.red600 : props.theme.color.gray700)};
  transition: color ${ANIMATION_DURATION} ease;

  ${(props) => props.$animate && animateOpacity()}
`;

const RadioGroupWrapper = styled(Card)<{ direction?: 'row' | 'column'; label?: string }>`
  display: flex;
  flex-direction: ${(props) => props.direction};
  & > * {
    flex: 1;
    ${(props) =>
      props.direction === 'row'
        ? `margin-left: ${props.theme.size.margin.m10};`
        : `margin-top: ${props.theme.size.margin.m10};`}
  }

  & > *:first-child {
    ${(props) => (props.direction === 'row' ? 'margin-left: 0px;' : 'margin-top: 0px;')}
  }
`;

const RadioGroup: FC<RadioGroupProps> = (props) => {
  const { label, tooltip, options, register, error, helperText, icon, ...newProps } = props;
  const didMountRef = useRef<boolean>(false);

  useEffect(() => {
    didMountRef.current = true;
  }, []);

  return (
    <Card display="flex" width="100%" flexDirection="column">
      <TopRow>
        {!!label && <Label>{label}</Label>}
        {!!tooltip && <Tooltip direction="top" text={tooltip} size="small" />}
      </TopRow>
      <RadioGroupWrapper {...newProps}>
        {options?.map((optionProps, index) => {
          const props = optionProps?.props ?? optionProps;
          return <Input key={index} register={register} {...props} />;
        })}
      </RadioGroupWrapper>
      {!!helperText && (
        <Card display="flex" alignItems="center" marginRight="4px">
          {icon && (
            <StyledIcon error={error} $animate={didMountRef.current} marginRight="4px">
              {CreateSVGComponent(error ? 'Error' : 'HelperInfo')}
            </StyledIcon>
          )}
          <HelperText data-testid="text-helper" error={error} $animate={didMountRef.current}>
            {helperText}
          </HelperText>
        </Card>
      )}
    </Card>
  );
};

export default RadioGroup;
