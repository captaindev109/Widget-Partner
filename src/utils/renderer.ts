import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
