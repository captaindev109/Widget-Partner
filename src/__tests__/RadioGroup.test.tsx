import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import RadioGroup from '@/components/RadioGroup';
import ApplicationThemeProvider from '@/providers/ApplicationThemeProvider';
import defaultTheme from '@/theme/theme';

/** Render tests */

const options = [
  {
    id: 'a',
    component: 'Input',
    type: 'radio',
    value: 'a',
    children: 'A',
  },
  {
    id: 'b',
    component: 'Input',
    type: 'radio',
    value: 'b',
    children: 'B',
  },
];

describe('RadioGroup render tests', () => {
  test('to see if the component render without crashing', () => {
    render(
      <ApplicationThemeProvider theme={defaultTheme}>
        <RadioGroup options={options} />
      </ApplicationThemeProvider>
    );
  });

  it('Snapshot test should pass', () => {
    const snap = renderer
      .create(
        <ApplicationThemeProvider theme={defaultTheme}>
          <RadioGroup options={options} />
        </ApplicationThemeProvider>
      )
      .toJSON();

    expect(snap).toMatchSnapshot();
  });
});
