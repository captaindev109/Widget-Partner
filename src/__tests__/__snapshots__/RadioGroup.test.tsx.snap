// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup render tests Snapshot test should pass 1`] = `
<div
  className="Card-sc-fb64gk-0 VEmTo"
  display="flex"
  width="100%"
>
  <div
    className="Card-sc-fb64gk-0 RadioGroup__TopRow-sc-1ihqkki-0 leCJCh hrJGZx"
  />
  <div
    className="Card-sc-fb64gk-0 RadioGroup__RadioGroupWrapper-sc-1ihqkki-3 leCJCh jBpDuX"
  >
    <div
      className="Card-sc-fb64gk-0 Input__RadioButtonContainer-sc-1g1qde7-5 leCJCh dIQPTq"
    >
      <input
        className="Input__BaseInput-sc-1g1qde7-0 Input__StyledRadioButton-sc-1g1qde7-9 loYxXz ijkpwp"
        data-testid="radio-input"
        id="a"
        type="radio"
        value="a"
      />
      <div
        className="Input__ContentWrapper-sc-1g1qde7-8 dPrdNA"
      >
        <div
          className="Card-sc-fb64gk-0 Input__RadioCircle-sc-1g1qde7-6 leCJCh jIHrWM"
        >
          <div
            className="Card-sc-fb64gk-0 Input__RadioIcon-sc-1g1qde7-7 leCJCh bbmyOl"
          >
            <svg
              fill="none"
              height="22px"
              pointerEvents="none"
              viewBox="0 0 23 22"
              width="22px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="#121212"
                height="22"
                rx="11"
                width="22"
                x="0.5"
              />
              <path
                d="M5.5 11L9.5 15L17.5 7"
                stroke="#FAFAFA"
                strokeWidth="2"
              />
            </svg>
          </div>
        </div>
        <div
          className="Card-sc-fb64gk-0 Input__ChildrenWrapper-sc-1g1qde7-11 gSNLuZ laAORz"
          width="100%"
        >
          A
        </div>
      </div>
    </div>
    <div
      className="Card-sc-fb64gk-0 Input__RadioButtonContainer-sc-1g1qde7-5 leCJCh dIQPTq"
    >
      <input
        className="Input__BaseInput-sc-1g1qde7-0 Input__StyledRadioButton-sc-1g1qde7-9 loYxXz ijkpwp"
        data-testid="radio-input"
        id="b"
        type="radio"
        value="b"
      />
      <div
        className="Input__ContentWrapper-sc-1g1qde7-8 dPrdNA"
      >
        <div
          className="Card-sc-fb64gk-0 Input__RadioCircle-sc-1g1qde7-6 leCJCh jIHrWM"
        >
          <div
            className="Card-sc-fb64gk-0 Input__RadioIcon-sc-1g1qde7-7 leCJCh bbmyOl"
          >
            <svg
              fill="none"
              height="22px"
              pointerEvents="none"
              viewBox="0 0 23 22"
              width="22px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="#121212"
                height="22"
                rx="11"
                width="22"
                x="0.5"
              />
              <path
                d="M5.5 11L9.5 15L17.5 7"
                stroke="#FAFAFA"
                strokeWidth="2"
              />
            </svg>
          </div>
        </div>
        <div
          className="Card-sc-fb64gk-0 Input__ChildrenWrapper-sc-1g1qde7-11 gSNLuZ laAORz"
          width="100%"
        >
          B
        </div>
      </div>
    </div>
  </div>
</div>
`;
